### PR TITLE
Added firmware checks

### PR DESCRIPTION
**Pull request by irritanterik**
https://github.com/matjaz/info.matjaz.candy/issues/5

### CODE REVIEW

File: lib/system.js [hunk 1]
```diff
@@ -49,4 +49,11 @@ function getSystemInfo () {
   })
 }
 
+function getUpdateInfo () {
+  return api('/manager/updates/update').then(result => {
+    return result
```
Comment (matjaz) [inline lib/system.js]: no need for then function

File: lib/systemFlows.js [hunk 2]
```diff
@@ -1,22 +1,41 @@
...
+  Homey.manager('flow').on('condition.systemFlowTimeout.units.autocomplete', onConditionSystemFlowTimeoutUnitsAutocomplete)
+  Homey.manager('flow').on('condition.systemNewFirmware', onConditionSystemNewFirmware)
+
+  checkActionSystemNewFirmwareTimeout = setInterval(checkActionSystemNewFirmware, checkTimeOut)
+  checkActionSystemNewFirmware()
+}
+
+function checkActionSystemNewFirmware () {
+  var tokens = {}
+  system.getUpdateInfo().then(function (updateInfo) {
```
Comment (matjaz) [inline lib/systemFlows.js:33]: use `Promise.all([system.getSystemInfo(), system.getUpdateInfo()])` for parallel requests

File: lib/systemFlows.js [hunk 3]
```diff
@@ -90,6 +109,12 @@ function onConditiononSystemFlowTimeout (callback, args) {
   callback(null, true)
 }
 
+function onConditionSystemNewFirmware (callback, args) {
+  system.getUpdateInfo().then(function (result) {
+    callback(null, (result.length && result[0].version))
```
Comment (matjaz) [inline lib/systemFlows.js]: condition callback second params should be boolean.
sth like `callback(null, !!(result.length && result[0].version))`

File: lib/systemFlows.js [hunk 4]
```diff
@@ -1,22 +1,41 @@
...
+function checkActionSystemNewFirmware () {
+  var tokens = {}
+  system.getUpdateInfo().then(function (updateInfo) {
+    if (!updateInfo.length || !updateInfo[0].version) return Promise.reject('no_update')
+    tokens.newVersion = updateInfo[0].version
+    return system.getSystemInfo()
+  }).then(function (systemInfo) {
+    tokens.currentVersion = systemInfo.homey_version
+    Homey.manager('flow').trigger('systemNewFirmware', tokens, null)
+    if (checkActionSystemNewFirmwareTimeout) clearTimeout(checkActionSystemNewFirmwareTimeout)
```
Comment (matjaz) [inline lib/systemFlows.js]: you're mixing `setInterval` with `clearTimeout`.
We want to be notified always for new versions, but only once for same version. So we need something as `latestTriggeredVersion` state, but trigger only when `newVersion !== currentVersion`.

**Review by matjaz — CHANGES_REQUESTED** [review]
thanks!

**Comment by matjaz** [comment]
also please resolve conflicts

File: lib/systemFlows.js (hunk 4 above)
Comment (matjaz) [reply lib/systemFlows.js]: we could cache `currentVersion` on boot. assuming reboot is always required.

**Comment by irritanterik** [comment]
Thanks! Bizarre clearTimeout works with a Interval :). Will push changes later this week.

File: lib/systemFlows.js (hunk 4 above)
Comment (irritanterik) [reply lib/systemFlows.js]: great suggestion

File: lib/systemFlows.js (hunk 3 above)
Comment (irritanterik) [reply lib/systemFlows.js]: Was not aware my code not resulted in a boolean. Will look into !!.

**Review by matjaz — APPROVED** [review]

**Comment by matjaz** [comment]
please resolve conflicts

**Comment by irritanterik** [comment]
Have them resolve. Because of DNS issues not able to sync them atm.

2016-10-21 20:33 GMT+02:00 Matjaž Lipuš notifications@github.com:

**Pull request merged**